### PR TITLE
8316342: CLHSDB "dumpclass" command produces invalid classes

### PR DESCRIPTION
**Pull request by ashu-mehra**
Please review this change to fix the operands of the bytecodes that operate on fields when dumping a class using SA.

Testing: hotspot_serviceability

I have also verified that `test/hotspot/jtreg/serviceability/sa/ClhsdbDumpclass.javaClhsdbDumpclass.java`, which is in the problem list because of this issue, passes with this change.
I have also verified it by dumping the class that has getstatic/putstatic bytecodes and comparing the bytecodes manually with the original classfile.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316342](https://bugs.openjdk.org/browse/JDK-8316342): CLHSDB "dumpclass" command produces invalid classes (**Bug** - P4)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15973/head:pull/15973` \
`$ git checkout pull/15973`

Update a local copy of the PR: \
`$ git checkout pull/15973` \
`$ git pull https://git.openjdk.org/jdk.git pull/15973/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15973`

View PR using the GUI difftool: \
`$ git pr show -t 15973`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15973.diff">https://git.openjdk.org/jdk/pull/15973.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15973#issuecomment-1740032272)

### CODE REVIEW

**Comment by plummercj** [comment]
@matias9927 and/or @coleenp Should have a look at this.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ByteCodeRewriter.java [hunk 1]
```diff
@@ -152,22 +116,29 @@ public void rewrite() {
...
+                    cpoolIndex = (short) cpCache.getEntryAt(cpci).getConstantPoolIndex();
                     writeShort(code, bci + 1, cpoolIndex);
                     break;
                 }
 
-                case Bytecodes._invokedynamic:
-                    cpoolIndex = getConstantPoolIndex(hotspotcode, bci + 1);
+                case Bytecodes._invokedynamic: {
+                    int cpci = method.getNativeIntArg(bci + 1);
+                    cpoolIndex = (short) cpCache.getIndyEntryAt(~cpci).getConstantPoolIndex();
```
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ByteCodeRewriter.java:137]: Is this really suppose to be `~cpci` and not just `cpci`?

Comment (ashu-mehra) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/tools/jcore/ByteCodeRewriter.java:137]: That's right. See https://github.com/openjdk/jdk/blob/ecb5e8a03f67c92d7956201de1fa7d07cc6af9cb/src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp#L1053 and https://github.com/openjdk/jdk/blob/ecb5e8a03f67c92d7956201de1fa7d07cc6af9cb/src/hotspot/share/oops/constantPool.hpp#L256

I tend to use `JvmtiClassFileReconstituter` as the reference for the code involved in dumping the classfile, and that makes things easier to follow and implement.

**Comment by plummercj** [comment]
Ok. The changes look good to me, but I don't have much of understanding of the hotspot code involved here, so I'll defer to a hotspot expert for the 2nd review.

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Nice simplification.
This looks good.
Thanks,
Serguei

**Comment by ashu-mehra** [comment]
I am returning from a break. Thanks @plummercj @sspitsyn for reviewing the changes.
It looks like this doesn't have any merge conflicts, so can still be integrated.

**Pull request closed without merging**